src/Sylius/Bundle/CoreBundle/Doctrine/ORM/ProductRepository.php

```diff
@@ -46,24 +47,18 @@ public function createListQueryBuilder($locale, $taxonId = null)
     /**
      * {@inheritdoc}
      */
-    public function createQueryBuilderByChannelAndTaxonSlug(ChannelInterface $channel, $taxonSlug, $locale)
+    public function createQueryBuilderByChannelAndTaxon(ChannelInterface $channel, TaxonInterface $taxon, $locale)
     {
         return $this->createQueryBuilder('o')
             ->addSelect('translation')
             ->leftJoin('o.translations', 'translation', 'WITH', 'translation.locale = :locale')
-            ->innerJoin('o.variants', 'variant')
-            ->innerJoin('variant.channelPricings', 'channelPricing')
```

Comment:
We can pass ``$criteria`` (so criteria array from URL) to this method and see if ``sorting[price]`` was set, if not - skip the joins.